tests/core/inference/test_device_manager.py

```diff
@@ -0,0 +1,90 @@
+import unittest
+from collections import OrderedDict
+from types import SimpleNamespace
+from unittest import mock
+
+import torch
+
+from marqo.core.exceptions import CudaDeviceNotAvailableError, CudaOutOfMemoryError
+from marqo.core.inference.device_manager import DeviceManager, Device
+
+
+class TestDeviceManager(unittest.TestCase):
+
+    def _device_manager_without_cuda(self):
+        with mock.patch("torch.cuda.is_available", return_value=False):
+            return DeviceManager()
+
+    def _device_manager_with_cuda(self, total_memory: int = 1_000_000):
+        with mock.patch("torch.cuda.is_available", return_value=True), \
+             mock.patch("torch.cuda.device_count", return_value=1), \
+             mock.patch("torch.cuda.get_device_name", return_value='cuda:0'), \
+             mock.patch("torch.cuda.get_device_properties", return_value=SimpleNamespace(total_memory=total_memory)):
+
+            return DeviceManager()
+
+    def test_init_with_cpu(self):
+        device_manager = self._device_manager_without_cuda()
+
+        self.assertEqual(device_manager.best_available_device_type, 'cpu')
+        self.assertEqual(device_manager.devices, [Device.cpu()])
+        self.assertFalse(device_manager._is_cuda_available_at_startup)
+
+    def test_init_with_gpu(self):
+        device_manager = self._device_manager_with_cuda(total_memory=1_000_000)
+
+        self.assertEqual(device_manager.best_available_device_type, 'cuda')
+        self.assertEqual(device_manager.devices, [Device.cpu(), Device.cuda(0, 'cuda:0', 1_000_000)])
+        self.assertTrue(device_manager._is_cuda_available_at_startup)
+
+    def test_cuda_health_check_should_skip_without_cuda_devices(self):
+        device_manager = self._device_manager_without_cuda()
+
+        with mock.patch("marqo.core.inference.device_manager.torch") as mock_cuda:
+            device_manager.cuda_device_health_check()
+        self.assertEqual(0, len(mock_cuda.mock_calls))
+
+    def test_cuda_health_check_should_pass_when_cuda_device_is_healthy(self):
```

Comment:
Let's add a check where there are multiple cuda devices. 1 or more is healthy, while some are unhealthy

Reply:
added